### ISSUE TITLE
[Feature request] Allow ask() to return dictionary of parameters when named

### ISSUE THREAD

**Issue by traviskaufman** [scikit-optimize/scikit-optimize#744]
When using the `ask()` interface with named parameters, e.g. `Integer(1, 10, name='num_things')`, it would be extremely convenient if there was a way to configure `ask` such that instead of returning an array, it returned a _dictionary_ of `{parameter_name: parameter_value}` entries. 

```python
SPACE = [
  Integer(1, 10, name='param_1'),
  Categorical(['A', 'B'], name='param_2'),
]

opt = Optimizer(SPACE)
opt.ask()  # Works as usual – [2, 'A']
opt.ask(use_named_args=True)  # {'param_1': 2, 'param_2': 'A'}
```

My use-case is I'm building a program which asks the user to perform some actions based on parameters and then has them report the results back. These actions take a long time so the program can't be running for the duration of it. Having `ask()` return the parameter names along with the values would allow me to decouple keeping track of the indices of each parameter in the part of the program where I ask the user to perform the action. 

Happy to open a PR for this if the team would consider it!

**Comment by traviskaufman** [comment]
friendly ping 😄

**Comment by iaroslav-ai** [comment]
Hi @traviskaufman ,
Would it work for you to use https://github.com/scikit-optimize/scikit-optimize/blob/e7d6618d214a5f242e3c5ca708bd347e498fa170/skopt/utils.py#L380 ?

**Comment by sotte** [comment]
From a usability standpoint it totally makes sense to have a `opt.ask(use_named_args=True)` or `opt.ask_named()` or something along the lines.

`point_asdict` can be a workaround. That said, I would assume the `point_asdict` also accepts a `Space` instance, but it's limited to dicts. It would not work for the scenario of @traviskaufman (he only has a list). `point_asdict` is also not sufficient when one loads the space from a yaml file (my scenario).
```python
space = Space.from_yaml("space.yaml")
opt = Optimizer(space)
opt.ask_named()  # Would be great!
```

**Comment by betatim** [comment]
Making `point_asdict()` also accept a space as first argument would be a good first step.

I wouldn't add a new method like `ask_named()` because it creates a weird API.

The best way forward IMHO is to improve `point_asdict()` to also work with a space instance and then make an example of a subclass of `Optimizer` that uses `point_asdict()` in `ask()` and its inverse in `tell()`.

The philosophy of scikit-optimize is to provide a simple ready made API via the `*_minimize` functions and then at a lower level of abstraction: all the building blocks you need to create a custom experience for exactly what you want to do. Instead of being a framework where you need to fit your code to how scikit-optimize thinks you should do things. I think providing the building bocks and examples of how to use them scales better than trying to support all possible and conflicting ways people want to use things.

**Comment by sotte** [comment]
I don't think that `ask_named()` would create a weird API, it would create a more obvious one IMHO. `opt.ask(use_named_args=True)` would also be more obvious than having to use `point_asdict()`. But I don't have a strong opinion on this :)

If we adjust `point_asdict()` then we probably have to adjust `point_aslist()` and `dimensions_aslist()` as well to accept `Space` as `search_space` to be consistent.

The documentation of `point_asdict` for the `point_as_list` parameter is a bit confusing. It states:
"The order of parameters in the list is given by sorted(params_space.keys())." Even after rereading it I'm not sure what it means. Does it expect the points to be already sorted by the keys of the space? Or is it telling me that the returned dict is going to be sorted?

If `point_asdict()` (and similar functions) accept `Space`, we would have to describe this `sorted` behaviour in case of dict and new behaviour for `Space`.

**Comment by betatim** [comment]
I think the next step is for someone to make a PR.

> "The order of parameters in the list is given by sorted(params_space.keys())." Even after rereading it I'm not sure what it means. Does it expect the points to be already sorted by the keys of the space? Or is it telling me that the returned dict is going to be sorted?

The problem that this function solves is to map a list like `[1., 2.234, 'a']` to a dictionary like `{'lr': 1., 'alpha': 2.234, 'foo': 'a'}. The assumption the function makes (and what that bit of docs is trying to say) is that it assumes that values in the list are sorted in the way that the search space dimension names are sorted. This means the resulting dictionary would actually be `{'lr': 'a', 'foo': 2.234, 'alpha': 1.}`

This bit of docs is about the parameter you are passing, so doesn't influence the return value (the default `dict` in Python doesn't support sorted keys).

**Comment by betatim** [comment]
My dislike of `ask_named()` is that it doesn't tell you what the method actually does. It would have to be something like `ask_with_named_dimensions()` or some such. The same with `used_named_args` which suggests that it relates to the arg(ument)s of the method but it doesn't. It is about the return value type.

This is why I think it is better to make a small class that derives from `Optimizer`, uses `point_as*` to do the conversion in `ask()` and `tell()` and has a good name so that people know what the class does.

**Comment by kabartay** [comment]
Seems the issue is still open.
